[MSE][GStreamer] Pause after seek is not working

calvaris opened the pull request:
#### e54b6b08878139eb460ad1094980014450c5e5b8
<pre>
[MSE][GStreamer] Pause after seek is not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=263317">https://bugs.webkit.org/show_bug.cgi?id=263317</a>

Reviewed by Philippe Normand.

When issuing seeks and play or pauses close to each other, there can be race conditions. To solve then first we need to
avoid changing the pipeline state while seeking. Then we need to ensure we know if the pipeline change was rejected or
failed, so the function signature was changed. We also need to consider that when a pause is requested by the element
and the player issues it to the pipeline, we are paused because it can take some time for the pipeline to pause and if
the element queries too fast, it can lead to more race conditions.

* LayoutTests/media/media-controller-play-then-pause-expected.txt:
* LayoutTests/media/media-controller-play-then-pause.html:
* LayoutTests/media/media-source/media-controller-media-source-play-then-pause-expected.txt: Added.
* LayoutTests/media/media-source/media-controller-media-source-play-then-pause.html: Added.
* LayoutTests/media/media-source/media-source-seek-and-play-expected.txt: Added.
* LayoutTests/media/media-source/media-source-seek-and-play.html: Added.
* LayoutTests/media/media-source/media-source-video-seek-pause-expected.txt: Added.
* LayoutTests/media/media-source/media-source-video-seek-pause.html: Added.
* LayoutTests/media/video-seek-pause-expected.txt: Added.
* LayoutTests/media/video-seek-pause.html: Added.
* LayoutTests/media/video-test.js:
(return.new.Promise):
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::isPipelineSeeking const):
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::seekToTarget):
(WebCore::MediaPlayerPrivateGStreamer::changePipelineState):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isPlayerPaused const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/273977@main">https://commits.webkit.org/273977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8ada138043776631c9514286d6e3b5097b913da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37178 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16079 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39749 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33180 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31673 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13643 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32727 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11819 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33366 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41006 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33747 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33740 "layout-tests (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37721 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12148 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9941 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35865 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13770 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32755 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8432 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12504 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13034 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -561,6 +575,8 @@ void MediaPlayerPrivateGStreamer::seekToTarget(const SeekTarget& inTarget)
     }
 
+    ASSERT(!m_isSeeking);
+    m_wasPausedBeforeSeeking = paused();
```
The field is named `m_wasPausedBeforeSeeking`, but if we arrive from the `// We can seek now.` we're setting this value *after* the seek, not before.

calvaris replied:
Well, I don't follow all seeka shananigans but if here we are setting `m_isSeeking = true`, we need to set `m_wasPausedBeforeSeeking` before that.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -441,6 +448,11 @@ bool MediaPlayerPrivateGStreamer::paused() const
 
+    if (m_isSeeking) {
+        GST_TRACE_OBJECT(pipeline(), "seeking, was paused before %s", boolForPrinting(m_wasPausedBeforeSeeking));
+        return m_wasPausedBeforeSeeking;
```
What does `paused()` return if the user calls .pause() after the seek?

calvaris replied:
I might not be understanding the question but after the seek `m_isSeeking` should be `false` and the answer to your question will depend on what is activated in the rest of the method.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -312,18 +314,26 @@ void MediaPlayerPrivateGStreamerMSE::sourceSetup(GstElement* sourceElement)
-        m_isPipelinePlaying = true;
-        m_playbackRatePausedState = PlaybackRatePausedState::Playing;
-    } else if (!shouldBePlaying && m_isPipelinePlaying) {
+            GST_DEBUG_OBJECT(pipeline(), "Setting the pipeline to PLAYING could not suceed right now");
```
In what cases?

calvaris replied:
Well, you can read that in the code yourself but I'll mention that in addition to failures, changing the pipeline while seeking will reject the change now as well. Pipeline will be moved to playing later after finishing the seek.

ntrrgc commented on LayoutTests/media/media-source/media-controller-media-source-play-then-pause.html:
```diff
@@ -0,0 +1,84 @@
+    await waitFor(controller, 'playing');
+
+	  testExpected('controller.playbackState', 'playing');
+    await sleepFor(2000);
```
There are some indentation issues here.

philn commented on LayoutTests/media/media-controller-play-then-pause.html:
```diff
@@ -39,10 +39,12 @@
 	consoleWrite("EVENT(play)");
 	testExpected('controller.playbackState', 'playing');
 	controller.removeEventListener('play', playAgain, true);
+  await sleepFor(2000);
```
2 seconds is a lot... The test runner might timeout more easily now. Was there a specific reason for waiting that much?

calvaris replied:
I wrote a new testing function for that. It works much better.

philn replied:
Seems to fail on iOS :(

calvaris replied:
Attempted fix.

philn commented on LayoutTests/media/media-source/media-controller-media-source-play-then-pause.html:
```diff
@@ -0,0 +1,84 @@
+    await waitFor(controller, 'playing');
+
+    testExpected('controller.playbackState', 'playing');
+    await sleepFor(2000);
```
Ditto

calvaris replied:
Done.

philn commented on LayoutTests/media/media-source/media-source-video-seek-pause.html:
```diff
@@ -0,0 +1,57 @@
+    run('video.currentTime = 2');
+    await sleepFor(5);
+    run('video.pause()');
+    await sleepFor(1000);
```
Can't we wait until the `paused` event is fired instead? Adding so many sleeps is going to be a source of flakyness...

calvaris replied:
Changed.

calvaris committed:
[MSE][GStreamer] Pause after seek is not working
https://bugs.webkit.org/show_bug.cgi?id=263317

Reviewed by Philippe Normand.

When issuing seeks and play or pauses close to each other, there can be race conditions. To solve then first we need to
avoid changing the pipeline state while seeking. Then we need to ensure we know if the pipeline change was rejected or
failed, so the function signature was changed. We also need to consider that when a pause is requested by the element
and the player issues it to the pipeline, we are paused because it can take some time for the pipeline to pause and if
the element queries too fast, it can lead to more race conditions.

* LayoutTests/media/media-controller-play-then-pause-expected.txt:
* LayoutTests/media/media-controller-play-then-pause.html:
* LayoutTests/media/media-source/media-controller-media-source-play-then-pause-expected.txt: Added.
* LayoutTests/media/media-source/media-controller-media-source-play-then-pause.html: Added.
* LayoutTests/media/media-source/media-source-seek-and-play-expected.txt: Added.
* LayoutTests/media/media-source/media-source-seek-and-play.html: Added.
* LayoutTests/media/media-source/media-source-video-seek-pause-expected.txt: Added.
* LayoutTests/media/media-source/media-source-video-seek-pause.html: Added.
* LayoutTests/media/video-seek-pause-expected.txt: Added.
* LayoutTests/media/video-seek-pause.html: Added.
* LayoutTests/media/video-test.js:
(return.new.Promise):
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::isPipelineSeeking const):
(WebCore::MediaPlayerPrivateGStreamer::play):
(WebCore::MediaPlayerPrivateGStreamer::pause):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::seekToTarget):
(WebCore::MediaPlayerPrivateGStreamer::changePipelineState):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isPlayerPaused const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/273977@main
